tests/helper.py

```diff
@@ -90,18 +90,29 @@ def assert_failed(self, with_errors=False, with_warnings=None):
             fail_on_warnings=with_warnings or with_warnings is None), \
                 "Test did not fail; failure was expected."
 
-        if with_errors:
-            assert self.err.errors, "Errors were expected."
-        elif self.err.errors:
-            raise AssertionError("Tests found unexpected errors: %s" %
-                                self.err.print_summary(verbose=True))
-
-        if with_warnings is not None:
-            if with_warnings:
-                assert self.err.warnings, "Warnings were expected."
-            elif self.err.warnings:
-                raise ("Tests found unexpected warnings: %s" %
-                           self.err.print_summary())
+        def check_message(message, props):
+            return all(message.get(prop) == value
+                       for prop, value in props.iteritems())
+
+        def find_message(messages, props):
+            return any(check_message(message, props)
+                       for message in messages)
+
+        def test_messages(mtype, expected):
+            messages = getattr(self.err, mtype)
+
+            if isinstance(expected, (list, tuple)):
+                assert all(find_message(messages, props)
+                           for props in expected)
+            elif expected:
+                assert messages, "Expected %s." % mtype
+            elif expected is not None:
```

Comment:
On Tue, May 05, 2015 at 07:05:49AM -0700, Mathieu Agopian wrote:
>Also, why not accept `None`? That would allow calling `test_messages('errors', None)` which is pretty clear? In that case the code would simply read `else:  # If expected is falsy.`.

Ah, sorry, I now see that I finished updating that docstring 
after I last pushed. I thought I'd done it before. For 
reference, then:

```python
        """
        Asserts that the error bundle has registered a failure. If
        `with_warnings` is any value any true value, or `None`, a warning
        is considered a failure.

        `with_warnings` or `with_errors` may be any of the following:

        * True: Messages of this type must be present.
        * False: Messages of this type must not be present.
        * None: Messages of this type may or may not be present.
        * Iterable of dicts: For dict returned by the iterator, at least
        one message must have a matching item for every key/value pair in the
        dict.
        """
```

(Note that this behavior, with relation to `True`/`False`/`None` 
is unchanged. I just made the behavior consistent between 
warnings and errors, and allowed specifying specific messages 
which were expected.)